pkg/tagset/threadsafe_factory.go

```diff
@@ -0,0 +1,102 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.Datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package tagset
+
+import "sync"
+
+// threadsafeFactory wraps another factory and uses a mutex to control
+// access.
+type threadsafeFactory struct {
+	sync.Mutex
+	Factory
+}
+
+// NewThreadsafeFactory wraps the given factory with a mutex, ensuring
+// thread-safe operation.
+func NewThreadsafeFactory(inner Factory) Factory {
+	return &threadsafeFactory{
+		Factory: inner,
+	}
+}
+
+// NewTags implements Factory.NewTags
+func (f *threadsafeFactory) NewTags(src []string) *Tags {
+	f.Lock()
+	tags := f.Factory.NewTags(src)
+	f.Unlock()
```

Comment:
Is there a reason to use this pattern instead of a `defer`? I know we prefer this to `defer f.Unlock()` in some situations but I am not sure why.

Reply:
I think `defer` still has a runtime cost in Go1.16 -- though I'm not sure where that's documented.  In this case, the usage is formulaic enough, and the control flow completely linear, that it's easy to avoid.  I've preferred defer in other cases.

Reply:
Correction: this is the _only_ use of sync.Mutex in this package.  I've preferred defer elsewhere in other PRs, though!

Reply:
https://github.com/golang/go/issues/14939

Reply:
A note that on such methods, with recent version of Go you could consider you fine. But I agree that `defer` should not always be use just because it exists, but we should use it when it's needed and when it is clarifying the reading, which is not the case here (but my personal opinion is bleeding in here).